### ISSUE TITLE
Vector3.slerp gives error in many situations and returns 0,0,0

### ISSUE THREAD

**Issue by nisovin** [godotengine/godot#27825]
**Godot version:** 3.1

**OS/device including version:** Win 10

**Issue description:**

When trying to use slerp in some (most?) circumstances, the console shows the error:

`Condition ' !p_axis.is_normalized() ' is true`

The function returns Vector3(0,0,0)

As far as I can tell, the implementation assumes that the cross-product of two normalized vectors is normalized, but that doesn't seem to be the case.

**Steps to reproduce:**

```
	var v1 = Vector3(0.5, 1, 0.5).normalized()
	var v2 = Vector3.DOWN
	var s = v1.slerp(v2, 0.5)
	print(s)
```

**Notes:** I admit I'm not a vector expert. If I have some fundamental misunderstanding of how slerp works, let me know.

**Comment by groud** [comment]
related #26363

**Comment by Toshiwoz** [comment]
Hello,
I've ran into a similar issue, not sure if I should create a different issue but, this is what I've found out:

In  godot/core/math/vector3.h  you have this:
https://github.com/godotengine/godot/blob/93845572d576133401d4f29b9bda1c90d261d577/core/math/vector3.h#L221-L228

And as you can see it uses the cross product within the formula, that is in the same file:
https://github.com/godotengine/godot/blob/93845572d576133401d4f29b9bda1c90d261d577/core/math/vector3.h#L158-L166

Problem seems to be that if you calculate the cross product without normalizing it, it doesn't always return a normalized vector, causing the slerp to fail.

I created a workaround in GDScript:

```
func slerp(va : Vector3, vb : Vector3, p_t : float):
	var theta = va.angle_to(vb)
	return va.rotated(va.cross(vb).normalized(), theta * p_t);
```

That works fine, I could fix it in C++ but I'm not sure if it's the most efficient way of doing that, computationally speaking.
If none thinks of a better solution in a few days I might create  a PR.

**Comment by whysguy52** [comment]
> Hello,
> I've ran into a similar issue, not sure if I should create a different issue but, this is what I've found out:
> 
> In godot/core/math/vector3.h you have this:
> https://github.com/godotengine/godot/blob/93845572d576133401d4f29b9bda1c90d261d577/core/math/vector3.h#L221-L228
> 
> And as you can see it uses the cross product within the formula, that is in the same file:
> https://github.com/godotengine/godot/blob/93845572d576133401d4f29b9bda1c90d261d577/core/math/vector3.h#L158-L166
> 
> Problem seems to be that if you calculate the cross product without normalizing it, it doesn't always return a normalized vector, causing the slerp to fail.
> 
> I created a workaround in GDScript:
> 
> ```
> func slerp(va : Vector3, vb : Vector3, p_t : float):
> 	var theta = va.angle_to(vb)
> 	return va.rotated(va.cross(vb).normalized(), theta * p_t);
> ```
> 
> That works fine, I could fix it in C++ but I'm not sure if it's the most efficient way of doing that, computationally speaking.
> If none thinks of a better solution in a few days I might create a PR.. ~as, you know I took my time to find out the problem, I'd like to get the credit for this, even if it's a very small thing.~
> Never mind, if I can do a pull with this change good, otherwise if someone else will be able to do it, good either.

So, If what you posted is actually the code in c++, the problem there would be, any 2 vectors that were parallel would yield a cross product of (0,0,0), which is a big problem. You can't normalize a zero vector! Your GD Script will also return a zero vector. What we should do is, check if the cross product == Vector3(0,0,0) and if it does, just return the original vector.  However, if the vectors are parallel but in opposite directions, you may want to return the second vector or something else since the original vector will obviously be pointing in the wrong direction. We should look into checking the cross and, if it's a zero vector, do something with the dot?

**Comment by Toshiwoz** [comment]
That's what I was referring to computationally efficient code, adding ifs into the function would make it not so efficient IMHO, anyhow I do understand it can't return an empty vector.

The reason I am tanking into account efficiency is that I've worked quite some time working on quadsphere heightmap projection for a game that I'll hopefully finish someday, and it uses slerp to generate the arc sphere of a planet. If this function becomes too slow it won't be able to render in real-time. I know, terrain generation is a niche thing, yet it'll be great if we could have a good set of interpolation functions.

I'm still dealing with a series of problems with my new pc, so I can't yet clone Godot, and thus not able to work on this for now.

Still, I'd like to help and I hope I'll solve my pc issue soon.

I would suggest we first discuss how to implement this, I've found this article about variants of the classical slerp that should be more performant:
https://zeux.io/2016/05/05/optimizing-slerp/

This code example in wikipedia seems to be closer to what I gues you're suggesting:
https://en.wikipedia.org/wiki/Slerp#Source_code
